design-proposals/klip-3-serialization-of-single-fields.md

```diff
@@ -0,0 +1,1076 @@
+  
+  Note: to be added as part of the future structured keys work.
+
+## Design
+
+### `CREATE STREAM` and `CREATE TABLE`
+
+C* statements will, by default, capture the values of 
+`ksql.persistence.wrap.single.keys` and `ksql.persistence.wrap.single.values`.
+
+Users can override these settings by providing any of 
+the `WRAP_SINGLE_XXX` family of `WITH` clause properties.
+
+These settings will be stored as part of the created source's metadata. 
+
+These settings control how the data in the source's topic should be
+deserialized by downstream queries.
+
+Providing any of these settings will cause an error to be returned should 
+the associated schema(s) be multi-field.   
+
+For example:
+
+```sql
+-- Default config: ksql.persistence.wrap.single.values=false
+
+-- creates a stream, picking up the system default of not wrapping
+-- the serialized value is expected to not be wrapped. 
+-- if the serialized value is wrapped it will likely result in a deserialization error.
+CREATE STREAM IMPLICIT_SOURCE (ID INT) WITH (...);
+
+-- override 'ksql.persistence.wrap.single.values' to true
+-- the serialized value is expected to not be unwrapped.
+CREATE STREAM EXPLICIT_SOURCE (ID INT) WITH (WRAP_SINGLE_VALUE=true, ...); 
+
+-- results in an error as the value schema is multi-field
+CREATE STREAM BAD_SOURCE (ID INT, NAME STRING) WITH (WRAP_SINGLE_VALUE=false, ...);
+
+-- [In future, with structured keys]
+-- override both settings to false.
+-- indicating that the key and value data in the topic are not wrapped
+CREATE TABLE SK_SOURCE (ID INT KEY, NAME STRING) WITH (WRAP_SINGLE_FIELDS=false, ...); 
+``` 
+
+### `CREATE STREAM AS` and `CREATE TABLE AS`
+
+C\*AS statements, by default, capture the values of 
+`ksql.persistence.wrap.single.keys` and `ksql.persistence.wrap.single.values`.
+
+Users can override these settings by providing any of 
+the `WRAP_SINGLE_XXX` family of `WITH` clause properties.
+
+These settings will be stored as part of the created source's metadata. 
+
+These settings control how the data in the source's topic should be
+serialized by this query and deserialized by any downstream queries.
+
+Providing any of these settings will cause an error to be returned should 
+the associated schema(s) be multi-field.    
+
+For example:
+
+```sql
+-- Default config: ksql.persistence.wrap.single.values=false
+
+-- creates a stream, picking up the system default of not wrapping
+-- the serialized values in the underlying topic will not be wrapped. 
+CREATE STREAM IMPLICIT_SOURCE AS SELECT ID FROM S;
+
+-- override 'ksql.persistence.wrap.single.values' to true
+-- the serialized values will be wrapped.
+CREATE STREAM EXPLICIT_SOURCE WITH(WRAP_SINGLE_VALUE=true) AS SELECT ID FROM S;
+
+-- results in an error as the value schema is multi-field
+CREATE STREAM BAD_SOURCE WITH(WRAP_SINGLE_VALUE=true) AS SELECT ID, COST FROM S;
+
+-- [In future, with structured keys]
+-- explicitly setting both settings to false.
```

Comment:
Not seeing where the "explicit setting" in this example is -- typo, perhaps?

Reply:
sortd.